File: CSSerializer/Test/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Buffer = CSSerializer.Buffer;

namespace Test
{
	class Game : CSSerializer.ISerializable
	{
		private int width, height;
		private Player player;
		private Player[] enemies;

		public Game()
		{
			width = 0;
			height = 0;
			player = null;
			enemies = null;
		}

		public Game(int width, int height, int numEnemies)
		{
			this.width = width;
			this.height = height;
			this.enemies = new Player[numEnemies];
			this.player = new Player(1);
			for (int i = 0; i < numEnemies; i++)
			{
				this.enemies[i] = new Player(i + 2);
			}
		}

		public static bool operator ==(Game g1, Game g2)
		{
			bool areEqual = g1.width == g2.width && g1.height == g2.height;
			if (!areEqual) return false;
			for (int i = 0; i < g1.enemies.Length; i++)
			{
				areEqual = areEqual && g1.enemies[i] == g2.enemies[i];
			}

			return areEqual && g1.player == g2.player;
		}

		public static bool operator !=(Game g1, Game g2)
		{
			return !(g1 == g2);
		}

		public void Serialize(Buffer buffer)
		{
			buffer.Write(width, height);
			buffer.Write(player);
			buffer.Write((short)enemies.Length);
			buffer.Write(enemies);
		}

		public void Deserialize(Buffer buffer)
		{
			buffer.Read(out width);
			buffer.Read(out height);
			buffer.Read(out player);
			short numEnemies;
			buffer.Read(out numEnemies);
			enemies = new Player[numEnemies];
			buffer.Read(enemies, numEnemies);
		}
	}
}

File: JBinBuff/test/Game.java
import com.binbuff.Buffer;
import com.binbuff.ISerializable;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Objects;

public class Game  implements ISerializable, Comparable<Game>, Serializable {

    private int width, height;
    private Player player;
    private Player[] enemies;

    public Game()
    {
        width = 0;
        height = 0;
        player = new Player();
        enemies = null;
    }

    public Game(int width, int height, int numEnemies)
    {
        this.width = width;
        this.height = height;
        this.enemies = new Player[numEnemies];
        this.player = new Player(1);
        for (int i = 0; i < numEnemies; i++)
        {
            this.enemies[i] = new Player(i + 2);
        }
    }

    @Override
    public void serialize(Buffer buffer) {
        buffer.write(width);
        buffer.write(height);
        buffer.write(player);
        buffer.write((short) enemies.length);
        buffer.write(enemies);
    }

    @Override
    public void deserialize(Buffer buffer) {
        this.width = buffer.readInt();
        this.height = buffer.readInt();
        buffer.read(player);
        short numEnemies = buffer.readShort();
        enemies = new Player[numEnemies];
        buffer.read(enemies);
    }

    @Override
    public boolean equals(Object o) {
        if(!(o instanceof Game)) return false;
        boolean res = true;
        Game other = (Game)o;
        res = this.width == other.width;
        res = res && this.height == other.height;
        res = res && this.player.equals(other.player);
        if (this.enemies.length != other.enemies.length) return false;
        for (int i = 0; i < this.enemies.length; i++) {
            res = res && this.enemies[i].equals(other.enemies[i]);
        }

        return res;
    }

    @Override
    public int compareTo(Game game) {
        return this.enemies.length - game.enemies.length;
    }


    @Override
    public int hashCode() {
        return Objects.hash(width, height, player, Arrays.hashCode(enemies));
    }
}
